ftw/solr/patches.py

```diff
@@ -0,0 +1,161 @@
+    doesn't matter here though, it's irrelevant how exactly they came to be).
+
+    In both of these cases, only the immediate parent of a subobject is
+    relevant for inheritance. Therefore, if an objects indexed security didn't
+    experience any changes, neither can any of its subobjects - recursively.
+
+    Because of this, downstream propagation can be stopped as soon as an
+    object is encountered whose indexed security didn't change.
+
+    ---
+
+    In the case of workflow changes, reindexObjectSecurity() still needs to
+    be called for every object that is *directly* affected by the change:
+
+    - switched to a different WF
+    - to a different WF state
+    - changes in the WF state's security settings
+    - object moved to/out of a placeful WF).
+
+    Identifying these objects is usually easy, and you can't and must not rely
+    on reinexObjectSecurity's recursion to pick them up. Instead, they can be
+    determined by querying for the relevant criteria, e.g. objects that have
+    a particular WF.
+
+    Recursion will then take care of updating security indexes for affected
+    subobjects that *don't* meet the criteria for being directly affected (
+    like subobjects with a different workflow).
+    """
+    indexes_to_update = []
+
+    # Since objectValues returns all objects, including placefulworkflow policy
+    # objects, we have to check if the object is Catalog aware.
+    if not ICatalogAware.providedBy(obj):
+        return
+
+    for index_name in obj._cmf_security_indexes:
+        if not is_index_up_to_date(catalog, obj, index_name):
+            indexes_to_update.append(index_name)
+
+    if len(indexes_to_update) > 0:
+        if not skip_self:
+            obj.reindexObject(idxs=indexes_to_update)
+
+        # We assume that if the parent is up to date, all children are too.
+        # Therefore we stop recursion as soon as we encounter an object
+        # whose security indexes are up to date.
+        for subobj in obj.objectValues():
+            recursive_index_security(catalog, subobj)
+
+
+def ftw_solr_CatalogMultiplex_reindexObjectSecurity(self, skip_self=False):
+    """Update security information in all registered catalogs.
+    """
+    if isFactoryContained(self):
+        return
+    at = getToolByName(self, TOOL_NAME, None)
+    if at is None:
+        return
+
+    catalogs = [c for c in at.getCatalogsByType(self.meta_type)
+                if ICatalogTool.providedBy(c)]
+
+    for catalog in catalogs:
+        recursive_index_security(catalog, self, skip_self=skip_self)
+
+
+def ftw_solr_CatalogAware_reindexObjectSecurity(self, skip_self=False):
+    """Reindex security-related indexes on the object.
+    """
+    catalog = self._getCatalogTool()
+    if catalog is None:
+        return
+
+    s = getattr(self, '_p_changed', 0)
+    recursive_index_security(catalog, self, skip_self=skip_self)
+    if s is None:
+        self._p_deactivate()
```

Comment:
That bit of logic is from the [original implementation of `reindexObjectSecurity` in `CMFCatalogAware`](https://github.com/zopefoundation/Products.CMFCore/blob/2.2.10/Products/CMFCore/CMFCatalogAware.py#L117-L120). My interpretation is that reindexing will unghost the object (since its state needs to be loaded to access attributes), and they want to re-ghost objects that previously were ghosts, in order to not have `reindexObjectSecurity` pollute the cache. Similar code is to be found in the [implementation of `ZopeFind()`](https://github.com/zopefoundation/Zope/blob/2.13.26/src/OFS/FindSupport.py#L108-L110).